### PR TITLE
ansible2: 2.2.0.0 -> 2.2.1.0

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

2.2.1.0 (as well as 2.1.4) fix a fairly nasty CVE: https://lwn.net/Articles/711357/

We do a few more things at the same time:

a) ansible is an application, not a package, so move it out of python-packages.nix
b) version 2.1 is an upstream supported version, so add that too
c) create a generic builder for versions 1.9, 2.1 and 2.2
d) fix ```kargo``` that depends on ```python2Packages.ansible2``` now that it has been moved out of ```pythonPackages```.

This PR covers both security fixes, new branch of code (2.1.x) as well as refactoring of the builder (less code, yeay!!) which may be OK for unstable but not necessarily ```16.09```.

Should we mark ansible 1 as broken as no fix has been released?

@grahamc, how would like me to proceed?

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] Linux
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
:+1: for moving packages out of `python-packages.nix`. We actually do this also for python modules now (so only the reference stays in `./pkgs/top-level/python-packages.nix` while the expressions move to `python-modules`)

File: pkgs/tools/admin/ansible/default.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    version = "2.1.4.0";
+    sha256  = "05nc68900qrzqp88970j2lmyvclgrjki66xavcpzyzamaqrh7wg9";
+  };
+
+  ansible2_2 = generic {
+    version = "2.2.1.0";
+    sha256  = "0gz9i30pdmkchi936ijy873k8di6fmf3v5rv551hxyf0hjkjx8b3";
+  };
+
+  ansible  = ansible1;
```
Comment (Mic92) [inline pkgs/tools/admin/ansible/default.nix]: When we do make to ansible2 the default ansible? Where did you find information regarding supported ansible versions?

**Comment by jgeerds** [comment]
@Mic92 I didn't know about this! Is there a official announcement / changelog ? I like that idea because python-packages.nix was/is bloated...

**Comment by grahamc** [comment]
(replying is on my to-do list for this afternoon)

**Comment by grahamc** [comment]
Yikes! Thank you!

Due to:

> a) ansible is an application, not a package, so move it out of python-packages.nix

I think we should separately patch stable and merge this PR (which I think is a nice improvement.)

> This PR covers both security fixes, new branch of code (2.1.x) as well as refactoring of the builder (less code, yeay!!) which may be OK for unstable but not necessarily 16.09.

Agreed, looks great :) thank you.

> Should we mark ansible 1 as broken as no fix has been released?

I don't _think_ so, not yet, let's hope they release an update soon?

Can you provide a patch for 16.09?

File: pkgs/tools/admin/ansible/default.nix (hunk 1 above)
Comment (grahamc) [inline pkgs/tools/admin/ansible/default.nix]: Agreed, should we make ansible2 default?

**Comment by FRidh** [comment]
> a) ansible is an application, not a package, so move it out of python-packages.nix
> d) fix kargo that depends on python2Packages.ansible2 now that it has been moved out of pythonPackages.

How does `kargo` use `ansible`? Does it import from it, or does it run it?

File: pkgs/tools/admin/ansible/default.nix (hunk 1 above)
Comment (globin) [inline pkgs/tools/admin/ansible/default.nix]: If you do so please add it to the 17.03 release notes

Comment (peterhoeg) [inline pkgs/tools/admin/ansible/default.nix]: I don't know of anything official from RedHat, but they specifically mentioned only releasing fixes for the 2.1 and 2.2 versions hence I'm assuming those are the only properly supported versions.

As for changing the default - that would only be in unstable, correct?

Comment (peterhoeg) [inline pkgs/tools/admin/ansible/default.nix]: @globin, I'll switch the default in a separate PR and include docs.

**Comment by peterhoeg** [comment]
@grahamc, shouldn't we just patch the version for 2.2 for stable instead of changing too many things?

@FRidh, ```kargo``` imports ```amsible```. Checking kargo after the change, the script runs and the import path gets set correctly.

**Comment by FRidh** [comment]
> @FRidh, kargo imports amsible. Checking kargo after the change, the script runs and the import path gets set correctly.

Exactly, so it is used as a library. I think that was also the reason why I put `ansible` in `python-packages.nix`, or at least, called it via `python-packages.nix`.

**Comment by peterhoeg** [comment]
> Exactly, so it is used as a library. I think that was also the reason why I put ansible in python-packages.nix, or at least, called it via python-packages.nix.

Fair enough. Maybe we should have a standard rule for this (further to @jgeerds comment) documented in the manual. My take on the current situation is:

If a piece of python software is mainly invoked on its own, it should go into its own file and be referenced from from ```all-packages.nix``` and use ```buildPythonApplication``` using a preferred python version (```python2Packages``` vs ```python3Packages```).

If instead it is primarily used as a dependency for another piece of python software, it should go into ```python-packages.nix``` and use ```buildPythonPackage``` to allow use with multiple python versions.

Everybody in agreement?

**Comment by FRidh** [comment]
@peterhoeg the problem is that as soon as you put it outside of `python-packages.nix` it is only available for a single Python interpreter version. We don't want to have libraries in `all-packages.nix` again, since that caused a lot of problems in the past. Therefore, I use the following rule myself:

> If it is supposed to be usable as a library, call it via `python-packages.nix`. Otherwise, put it elsewhere.

I don't know whether `ansible` is supposed to be usable as a library, but clearly it is used like that.

**Comment by peterhoeg** [comment]
@FRidh, can't most python packages be used as libraries? In any case, I'll change that part so it goes back to the ```python-packages.nix``` but uses the generric builder instead.

**Comment by fpletz** [comment]
Ansible upstream won't patch the 1.x versions because it's too complicated. These versions should be removed from master and 16.09 (or at least mark at broken for 16.09).

**Comment by globin** [comment]
Python applications should IMHO be referenced from python-packages.nix and then be aliased in all-packages.nix with e.g `ansible = python3Packages.ansible`

**Comment by FRidh** [comment]
@FRidh, can't most python packages be used as libraries? In any case, I'll change that part so it goes back to the python-packages.nix but uses the generric builder instead.

If they install their modules in `site-packages`, then yes. But that's not a requirement for applications, only for libraries.

**Comment by globin** [comment]
> If they install their modules in site-packages, then yes. But that's not a requirement for applications, only for libraries.

That means everything using `buildPython{Package,Application}`, right?

**Comment by FRidh** [comment]
> That means everything using buildPython{Package,Application}, right?

Indeed, everything using that function puts modules in `site-packages` and is thus usable as a library. 
( That was also the issue we had in the past with `prefix PYTHONPATH` instead of `set PYTHONPATH`; if a Python application uses a Python application you don't want to merge their `PYTHONPATH`. )

**Comment by grahamc** [comment]
@FRidh what is your call on this? Should we merge? Should we not merge?

**Comment by peterhoeg** [comment]
@fridh and @grahamc, I will not have time to do it until tonight but I'll make the following changes to this PR:

 - move the definitions back to python-packages.nix
 - drop ansible1
 - make 2.2 the default
 - update the documentation

Everybody cool?

The update to 2.2.1.0 has already been applied to stable.

**Comment by grahamc** [comment]
👍 thank you!

**Comment by FRidh** [comment]
@peterhoeg yep, go ahead. 

Just remember that whenever you change from `buildPythonPackage` to `buildPythonApplication` you change the name of the package and thus break upgrading with `nix-env -u`.

**Comment by peterhoeg** [comment]
All done @FRidh and @grahamc.

**Review by grahamc — APPROVED** [review]
@FRidh LGTY? If you don't leave feedback in ~24hrs I'll go ahead and merge.

**Comment by FRidh** [comment]
I've pushed 377b05ae38ef6c81d83e0c01446a69ea8fe287d2, 46b1ea260a05d2d5fb01624c93d405daaf750b0c and c42cfa1e91bf271913aaa055ab34b2416df5b2f8.

While there is convenience in having a generic function, it becomes very unflexible as soon as versions start to diverge.

**Pull request closed without merging**